modules/service/src/main/resources/db/migration/V0371__target_role.sql

```diff
@@ -0,0 +1,66 @@
+create type e_target_role as enum('science', 'guide');
+comment on type e_existence is 'The role of the target.';
+
+alter table t_target
+add column c_role e_target_role;
+
+update t_target
+set c_role = 'science';
+
+alter table t_target
+alter column c_role set not null;
+
+CREATE OR REPLACE FUNCTION asterism_target_is_science(d_target_id)
+  RETURNS BOOLEAN AS $$
+DECLARE
+  role e_target_role;
+BEGIN
+  SELECT c_role
+  INTO role
+  FROM t_target
+  WHERE c_target_id = $1;
+  IF FOUND AND (role != 'science') THEN
+    RAISE EXCEPTION 'Asterism target must be a science target.';
+  ELSE
+    RETURN TRUE;
+  END IF;
+END;
+$$ LANGUAGE plpgsql;
+
+alter table t_asterism_target 
+add constraint asterism_target_is_science check (asterism_target_is_science(c_target_id));
+
+-- The above prevents assigning a non-science target to an asterism, but doesn't protect
+-- against the role of a target changing.
+CREATE OR REPLACE FUNCTION target_role_immutable()
+  RETURNS trigger AS $$
+DECLARE
+BEGIN
+  IF NEW.c_role != OLD.c_role THEN
+    RAISE EXCEPTION 'The role of a target cannot be changed.';
+  END IF;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER t_target_role_immutable_trigger
+BEFORE UPDATE on t_target
+FOR EACH ROW
+EXECUTE FUNCTION target_role_immutable();
+
+-- Need to recreate target view because of added field in target table.
+-- Can't use `update_table_with_views` because the saved view definition has
+-- expanded the `*` to the explicit columns, so the new column won't get added.
```

Comment:
Could you expand on this, I'm not sure I understand.  I guess maybe I expected the view definition itself to be stored and not the result of executing the view definition.  It sounds like that's not the case?

Reply:
I expected the same, but apparently the stored definition has already expanded the `*`. The output of 

> select pg_get_viewdef('v-target`, true);

is
```sql
 SELECT t_target.c_program_id,
    t_target.c_target_id,
    t_target.c_name,
    t_target.c_existence,
    t_target.c_type,
    t_target.c_sid_ra,
    t_target.c_sid_dec,
    t_target.c_sid_epoch,
    t_target.c_sid_pm_ra,
    t_target.c_sid_pm_dec,
    t_target.c_sid_rv,
    t_target.c_sid_parallax,
    t_target.c_sid_catalog_name,
    t_target.c_sid_catalog_id,
    t_target.c_sid_catalog_object_type,
    t_target.c_nsid_des,
    t_target.c_nsid_key_type,
    t_target.c_nsid_key,
    t_target.c_source_profile,
    t_target.c_role,
        CASE
            WHEN t_target.c_type = 'sidereal'::e_target_type THEN t_target.c_target_id::character varying
            ELSE NULL::character varying
        END AS c_sidereal_id,
        CASE
            WHEN t_target.c_type = 'nonsidereal'::e_target_type THEN t_target.c_target_id::character varying
            ELSE NULL::character varying
        END AS c_nonsidereal_id,
        CASE
            WHEN t_target.c_sid_catalog_name IS NOT NULL THEN t_target.c_target_id::character varying
            ELSE NULL::character varying
        END AS c_sid_catalog_info_id,
        CASE
            WHEN t_target.c_sid_pm_ra IS NOT NULL THEN t_target.c_target_id::character varying
            ELSE NULL::character varying
        END AS c_sid_pm_id,
        CASE
            WHEN t_target.c_sid_parallax IS NOT NULL THEN t_target.c_target_id::character varying
            ELSE NULL::character varying
        END AS c_sid_parallax_id,
        CASE
            WHEN t_target.c_sid_rv IS NOT NULL THEN t_target.c_target_id::character varying
            ELSE NULL::character varying
        END AS c_sid_rv_id
   FROM t_target;
```

Reply:
Perhaps then `update_table_with_views` isn't very useful.

Reply:
`update_table_with_views` works well if you modify a column, just not if you add or delete a column